sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1802,3 +1803,85 @@ object OptimizeLimitZero extends Rule[LogicalPlan] {
       empty(ll)
   }
 }
+
+/**
+ * Removes Cast expressions from predicates to push down the filters. This will allow filter push
+ * down for byte and short columns without explicitly casting literals.
```

Comment:
So, what you're doing right now is basically https://github.com/apache/spark/pull/8718 which was rejected.

Reply:
In this solution i am not generalising casting. This PR only deals with Byte and Short columns and Values are only downcasted in case they are within byte/short range. Otherwise whole Predicate is left unchanged. 
Also before pushing i am explicitly checking whether these values can be pushed down or not.

Reply:
The approach basically looks same. Also, we should better generalise it so other data sources can support. The current approach seems difficult to extend.